cabal-install/Distribution/Client/CmdInstall/Utils.hs

```diff
@@ -0,0 +1,41 @@
+module Distribution.Client.CmdInstall.Utils
+  ( reportCannotPruneDependencies
+  , warnIfNoExes
+  ) where
+
+import Prelude ()
+import Distribution.Client.Compat.Prelude
+
+import qualified Data.Map as Map
+
+import Distribution.Client.CmdErrorMessages
+       ( plural, listPlural, renderCannotPruneDependencies, showTargetSelector )
+import Distribution.Client.ProjectOrchestration
+       ( CannotPruneDependencies(..), ComponentName(..)
+       , ComponentTarget(..) , ProjectBuildContext(..) )
+import Distribution.Simple.Utils
+       ( die', warn )
+import Distribution.Verbosity
+       ( Verbosity )
+
+-- | Emits a warning if the build context does not contain any executables.
+warnIfNoExes :: Verbosity -> ProjectBuildContext -> IO ()
+warnIfNoExes verbosity buildCtx =
+  when noExes $
+    warn verbosity $
+    "You asked to install executables, but there are no executables in "
+    <> plural (listPlural selectors) "target" "targets" <> ": "
+    <> intercalate ", " (showTargetSelector <$> selectors) <> ". "
+    <> "Perhaps you want to use --lib to install libraries instead."
+  where
+    targets    = concat $ Map.elems $ targetsMap buildCtx
+    components = fst <$> targets
+    selectors  = concatMap snd targets
+    noExes     = null $ catMaybes $ exeMaybe <$> components
+
+    exeMaybe (ComponentTarget (CExeName exe) _) = Just exe
+    exeMaybe _                                  = Nothing
+
+reportCannotPruneDependencies :: Verbosity -> CannotPruneDependencies -> IO a
```

Comment:
On the other hand, this seems to be used in both `CmdBuild` and `CmdInstall` so having it in the module with `CmdInstall` in the name doesn't feel right.

Reply:
Yes, that's fine. Please don't create `.Utils` modules though. They are *impossible* to get rid off.

Reply:
I don't know if I'd say _impossible_ ;) In any case renamed both of them to `Internal`.